src/vs/editor/common/modes/textToHtmlTokenizer.ts

```diff
@@ -8,11 +8,115 @@ import * as strings from 'vs/base/common/strings';
+import { CharacterMapping } from 'vs/editor/common/viewLayout/viewLineRenderer';
+import { CharCode } from 'vs/base/common/charCode';
+import { ViewLineToken } from 'vs/editor/common/core/viewLineToken';
 
 export function tokenizeToString(text: string, languageId: string): string {
 	return _tokenizeToString(text, _getSafeTokenizationSupport(languageId));
 }
 
+export function tokenizeLineToHTML(text: string, viewLineTokens: ViewLineToken[], rules: { [key: string]: string }, options: { startOffset: number, endOffset: number, tabSize: number, containsRTL: boolean }): string {
+	let tabSize = options.tabSize;
+	let containsRTL = options.containsRTL;
+
+	let result = `<div>`;
+	const characterMapping = new CharacterMapping(text.length + 1, viewLineTokens.length);
+
+	let charIndex = options.startOffset;
+	let tabsCharDelta = 0;
+	let charOffsetInPart = 0;
+
+	for (let tokenIndex = 0, lenJ = viewLineTokens.length; tokenIndex < lenJ; tokenIndex++) {
+		const token = viewLineTokens[tokenIndex];
+		const tokenEndIndex = token.endIndex;
+
+		if (token.endIndex < options.startOffset) {
+			continue;
+		}
+
+		const tokenType = token.type;
+		let partContentCnt = 0;
+		let partContent = '';
+
+		for (; charIndex < tokenEndIndex && charIndex < options.endOffset; charIndex++) {
+			characterMapping.setPartData(charIndex, tokenIndex, charOffsetInPart);
+			const charCode = text.charCodeAt(charIndex);
+
+			switch (charCode) {
+				case CharCode.Tab:
+					let insertSpacesCount = tabSize - (charIndex + tabsCharDelta) % tabSize;
+					tabsCharDelta += insertSpacesCount - 1;
+					charOffsetInPart += insertSpacesCount - 1;
+					while (insertSpacesCount > 0) {
+						partContent += '&nbsp;';
+						partContentCnt++;
+						insertSpacesCount--;
+					}
+					break;
+
+				case CharCode.Space:
+					partContent += '&nbsp;';
+					partContentCnt++;
+					break;
+
+				case CharCode.LessThan:
+					partContent += '&lt;';
+					partContentCnt++;
+					break;
+
+				case CharCode.GreaterThan:
+					partContent += '&gt;';
+					partContentCnt++;
+					break;
+
+				case CharCode.Ampersand:
+					partContent += '&amp;';
+					partContentCnt++;
+					break;
+
+				case CharCode.Null:
+					partContent += '&#00;';
+					partContentCnt++;
+					break;
+
+				case CharCode.UTF8_BOM:
+				case CharCode.LINE_SEPARATOR_2028:
+					partContent += '\ufffd';
+					partContentCnt++;
+					break;
+
+				case CharCode.CarriageReturn:
+					// zero width space, because carriage return would introduce a line break
+					partContent += '&#8203';
+					partContentCnt++;
+					break;
+
+				default:
+					partContent += String.fromCharCode(charCode);
+					partContentCnt++;
+			}
+
+			charOffsetInPart++;
+		}
+
+		characterMapping.setPartLength(tokenIndex, partContentCnt);
+		let style = tokenType.split(' ').map(type => rules[type]).join('');
+		if (containsRTL) {
+			result += `<span dir="ltr" style="${style}">${partContent}</span>`;
```

Comment:
perhaps this is overkill. I have a feeling the software where we paste should figure out by itself what to do with RTL characters, so I suggest to ignore RTL here for now.